test/kubernetes_unittest.cc

```diff
@@ -696,4 +699,137 @@ TEST_F(KubernetesTest, GetPodAndContainerMetadata) {
+  for (int i = 0; i < 30; i++){
+    const auto metadata_map = store.GetMetadataMap();
+    const auto m = metadata_map.find(resource);
+    if (m != metadata_map.end() && m->second.collected_at > last_timestamp) {
+      return true;
+    }
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+  return false;
+}
+
+TEST_F(KubernetesTest, KubernetesUpdater) {
+  // Create a fake server representing the Kubernetes master.
+  testing::FakeServer server;
+  server.SetResponse("/api/v1/nodes?limit=1", "{}");
+  server.SetResponse("/api/v1/pods?limit=1", "{}");
+  server.AllowStream(
+      "/api/v1/pods?fieldSelector=spec.nodeName%3DTestNodeName&watch=true");
+  server.AllowStream(
+      "/api/v1/watch/nodes/TestNodeName?watch=true");
+
+  // Start the updater with a config that points to the fake server.
+  Configuration config(std::istringstream(
+      "InstanceId: TestID\n"
+      "InstanceResourceType: gce_instance\n"
+      "InstanceZone: TestZone\n"
+      "KubernetesClusterLocation: TestClusterLocation\n"
+      "KubernetesClusterName: TestClusterName\n"
+      "KubernetesEndpointHost: " + server.GetUrl() + "\n"
+      "KubernetesNodeName: TestNodeName\n"
+      "KubernetesUseWatch: true\n"
+  ));
+  MetadataStore store(config);
+  KubernetesUpdater updater(config, /*health_checker=*/nullptr, &store);
+  std::thread updater_thread([&updater] { updater.start(); });
+
+  // Wait for updater's watchers to connect to the server (hanging GETs).
+  server.WaitForOneStreamWatcher(
+      "/api/v1/pods?fieldSelector=spec.nodeName%3DTestNodeName&watch=true");
+  server.WaitForOneStreamWatcher(
+      "/api/v1/watch/nodes/TestNodeName?watch=true");
+
+  // For nodes, send stream responses from the fake Kubernetes
+  // master and verify that the updater propagates them to the store.
+  Timestamp last_nodes_timestamp = std::chrono::system_clock::now();
+  for (int i = 0; i < 3; i++) {
+    json::value resp = json::object({
+      {"type", json::string("ADDED")},
+      {"object", json::object({
+        {"metadata", json::object({
+          {"name", json::string("TestNodeName")},
+          {"creationTimestamp", json::string("2018-03-03T01:23:45.678901234Z")},
+        })}
+      })}
+    });
+    // Send a response to the watcher.
+    server.SendStreamResponse(
+        "/api/v1/watch/nodes/TestNodeName?watch=true",
+        resp->ToString());
+    // Wait until watcher has processed response (by polling the store).
+    MonitoredResource resource("k8s_node",
+                               {{"cluster_name", "TestClusterName"},
+                                {"location", "TestClusterLocation"},
+                                {"node_name", "TestNodeName"}});
+    EXPECT_TRUE(
+        WaitForNewerCollectionTimestamp(store, resource, last_nodes_timestamp));
+
+    const auto metadata_map = store.GetMetadataMap();
+    const auto& metadata = metadata_map.at(resource);
```

Comment:
Do we need to add an assertion here to ensure metadata is not nil?

Reply:
`WaitForNewerCollectionTimestamp` guarantees that the key will be in the map. And a reference cannot be null in C++.